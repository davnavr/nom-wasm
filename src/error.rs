//! Contains types describing errors that occur during parsing.

mod add_cause;
mod cause;
mod component;
mod invalid_flags;
mod invalid_tag;
mod length_mismatch;

#[cfg(feature = "alloc")]
#[cfg_attr(doc_cfg, doc(cfg(feature = "alloc")))]
mod verbose_error;

#[doc(no_inline)]
pub use crate::{
    isa::{InvalidExpr, InvalidInstr, InvalidOpcode},
    module::preamble::InvalidMagic,
    values::InvalidVector,
};

pub use cause::ErrorCause;
pub use component::{ImportComponent, LimitsComponent, MemArgComponent};
pub use invalid_flags::{InvalidFlags, InvalidFlagsValue};
pub use invalid_tag::InvalidTag;
pub use length_mismatch::LengthMismatch;

#[cfg(feature = "alloc")]
pub use verbose_error::VerboseError;

pub(crate) use add_cause::AddCause;

/// Default error type, which tracks an error's location and the reason why it occured.
#[derive(Clone, Eq, PartialEq)]
#[non_exhaustive]
pub struct Error<'a> {
    /// A slice into the original input indicating where the error occured.
    pub input: &'a [u8],
    /// An error code indicating why a parse failed.
    pub cause: ErrorCause,
}

impl core::fmt::Debug for Error<'_> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_struct("Error")
            .field("input", &crate::hex::Bytes(self.input))
            .field("cause", &self.cause)
            .finish()
    }
}

impl<'a> From<nom::error::Error<&'a [u8]>> for Error<'a> {
    #[inline]
    fn from(error: nom::error::Error<&'a [u8]>) -> Self {
        Self {
            input: error.input,
            cause: ErrorCause::Nom(error.code),
        }
    }
}

impl<'a> nom::error::ParseError<&'a [u8]> for Error<'a> {
    #[inline]
    fn from_error_kind(input: &'a [u8], kind: nom::error::ErrorKind) -> Self {
        Self {
            input,
            cause: ErrorCause::Nom(kind),
        }
    }

    #[inline]
    fn append(_: &'a [u8], _: nom::error::ErrorKind, other: Self) -> Self {
        other
    }
}

impl core::fmt::Display for Error<'_> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        core::fmt::Display::fmt(&self.cause, f)?;

        if !self.input.is_empty() {
            write!(f, ", for {} bytes in input:", self.input.len())?;

            const DISPLAY_MAX: usize = 8;

            for b in self.input.iter().take(DISPLAY_MAX) {
                write!(f, " {b:02X}")?;
            }

            if self.input.len() > DISPLAY_MAX {
                f.write_str(" ...")?;
            }
        }

        Ok(())
    }
}

#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
#[cfg(feature = "std")]
impl std::error::Error for Error<'_> {
    #[inline]
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        std::error::Error::source(&self.cause)
    }
}

/// Trait for error types used with [`nom-wasm`](crate).
pub trait ErrorSource<'a>: nom::error::ParseError<&'a [u8]> {
    /// Combines existing error with a newly constructed error.
    #[inline]
    fn append_with_cause(input: &'a [u8], cause: ErrorCause, other: Self) -> Self {
        let _ = (input, cause);
        other
    }

    /// Creates a new error from the input where it occured, and its [`ErrorCause`].
    #[inline]
    fn from_error_cause(input: &'a [u8], cause: ErrorCause) -> Self {
        Self::from_error_kind(input, cause.to_error_kind())
    }
}

impl ErrorSource<'_> for () {}

impl<'a> ErrorSource<'a> for (&'a [u8], nom::error::ErrorKind) {}

impl<'a> ErrorSource<'a> for nom::error::Error<&'a [u8]> {}

impl<'a> ErrorSource<'a> for nom::error::VerboseError<&'a [u8]> {}

impl<'a> ErrorSource<'a> for Error<'a> {
    #[inline]
    fn from_error_cause(input: &'a [u8], cause: ErrorCause) -> Self {
        Self { input, cause }
    }
}
