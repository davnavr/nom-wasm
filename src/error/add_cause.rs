use crate::{
    error::{ErrorCause, ErrorSource},
    input::Result,
};

pub(crate) trait AddCause<'a, T, E: ErrorSource<'a>> {
    fn add_cause_with<F: FnOnce() -> ErrorCause>(self, f: F) -> Self;

    fn add_cause(self, cause: ErrorCause) -> Self;
}

impl<'a, T, E: ErrorSource<'a>> AddCause<'a, T, E> for Result<T, E> {
    #[inline]
    fn add_cause_with<F: FnOnce() -> ErrorCause>(self, f: F) -> Self {
        match self {
            Ok(value) => Ok(value),
            Err(err) => Err(err.map(|e| e.with_cause(f()))),
        }
    }

    #[inline]
    fn add_cause(self, cause: ErrorCause) -> Self {
        match self {
            Ok(value) => Ok(value),
            Err(err) => Err(err.map(|e| e.with_cause(cause))),
        }
    }
}
