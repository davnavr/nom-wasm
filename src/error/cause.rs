use core::fmt::{Display, Formatter};

/// Describes an [`ErrorCause`] where the length of some data was incorrect.
#[derive(Clone, Debug, Eq, PartialEq)]
#[allow(missing_docs)]
pub struct LengthMismatch {
    pub expected: u32,
    pub actual: u32,
}

impl LengthMismatch {
    fn print(&self, name: &str, f: &mut Formatter) -> core::fmt::Result {
        write!(
            f,
            "expected {} bytes for {name}, but got {}",
            self.expected, self.actual
        )
    }
}

/// Error type used when a byte or 32-bit enumeration value was invalid.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[non_exhaustive]
pub enum InvalidTag {
    /// An invalid [`ModuleSectionId`](crate::module::ModuleSectionId).
    ModuleSectionId(u8),
    #[allow(missing_docs)]
    FuncType(Option<u8>),
    /// An invalid [`ImportDesc`](crate::module::ImportDesc).
    ImportDesc(Option<u8>),
}

impl Display for InvalidTag {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let (value, value_width) = match self {
            Self::ModuleSectionId(b) => (Some(u32::from(*b)), 4),
            Self::FuncType(b) | Self::ImportDesc(b) => (b.map(u32::from), 4),
        };

        let name = match self {
            Self::ModuleSectionId(_) => "module section ID",
            Self::FuncType(_) => "function type",
            Self::ImportDesc(_) => "import desc",
        };

        if let Some(value) = value {
            write!(
                f,
                "the {name} tag {value:#0value_width$X} ({value}) is invalid"
            )
        } else {
            write!(f, "missing {name} tag")
        }
    }
}

#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
#[cfg(feature = "std")]
impl std::error::Error for InvalidTag {}

/// Used with [`InvalidFlags`] to indicate what values were invalid.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[allow(clippy::exhaustive_enums)]
pub enum InvalidFlagsValue<V: Copy> {
    /// The flags were parsed, but had invalid flags.
    Invalid {
        /// The flags value that *contains* the invalid flags.
        value: V,
        /// The flags that caused validation to fail.
        invalid: V,
    },
    /// Flags could not be parsed.
    Missing,
}

/// Error type used when a byte or 32-bit flags combination was invalid.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[non_exhaustive]
pub enum InvalidFlags {
    /// Invalid flags for [`Limits`](crate::types::Limits).
    Limits(InvalidFlagsValue<u8>),
    /// Invalid flags for a [`GlobalType`](crate::types::GlobalType).
    GlobalType(InvalidFlagsValue<u8>),
}

impl Display for InvalidFlags {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        let (name, invalid) = match self {
            Self::Limits(e) => ("limits", e),
            Self::GlobalType(e) => ("global type", e),
        };

        match invalid {
            InvalidFlagsValue::Invalid { value, invalid } => write!(
                f,
                "the {name} flags {value:#04X} contains invalid flag(s): {invalid:#04X}"
            ),
            InvalidFlagsValue::Missing => write!(f, "missing {name} flags"),
        }
    }
}

#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
#[cfg(feature = "std")]
impl std::error::Error for InvalidFlags {}

/// Indicates which part of a [`Limits`](crate::types::Limits) could not be parsed.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[non_exhaustive]
#[allow(missing_docs)]
pub enum LimitsComponent {
    Minimum,
    Maximum,
}

impl Display for LimitsComponent {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.write_str(match self {
            Self::Minimum => "minimum",
            Self::Maximum => "maximum",
        })
    }
}

/// Indicates which field of an [`Import`](crate::module::Import) could not be parsed.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[non_exhaustive]
#[allow(missing_docs)]
pub enum ImportComponent {
    Module,
    Name,
}

impl Display for ImportComponent {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.write_str(match self {
            Self::Module => "module name",
            Self::Name => "import name",
        })
    }
}

/// Indicates why a [`MemArg`](crate::isa::MemArg) could not be parsed.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[non_exhaustive]
#[allow(missing_docs)]
pub enum MemArgComponent {
    /// Indicates that the [**`align`**] field could not be parsed when [`None`]; otherwise,
    /// indicates that the [**`align`**] was too large.
    ///
    /// [**`align`**]: crate::isa::MemArg::align
    Alignment(Option<u32>),
    Offset,
    Memory,
}

impl Display for MemArgComponent {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Alignment(Some(a)) => {
                write!(f, "specified alignment was 2^{a}, which is too large")
            }
            Self::Alignment(None) => f.write_str("alignment field"),
            Self::Offset => f.write_str("offset field"),
            Self::Memory => f.write_str("memory field"),
        }
    }
}

/// Describes why a parser error occured.
#[derive(Clone, Debug, Eq, PartialEq)]
#[non_exhaustive]
#[allow(missing_docs)]
pub enum ErrorCause {
    Nom(nom::error::ErrorKind),
    Leb128 {
        destination: crate::values::leb128::Destination,
        reason: crate::values::leb128::InvalidEncoding,
    },
    InvalidTag(InvalidTag),
    InvalidFlags(InvalidFlags),
    Vector(crate::values::InvalidVector),
    #[non_exhaustive]
    NameLength,
    NameContents(LengthMismatch),
    NameEncoding(core::str::Utf8Error),
    Index(&'static &'static str),
    #[non_exhaustive]
    SectionId,
    #[non_exhaustive]
    SectionLength,
    SectionContents(LengthMismatch),
    #[non_exhaustive]
    CustomSectionName,
    PreambleMagic(crate::module::preamble::InvalidMagic),
    PreambleVersion(Option<u32>),
    /// A [`BlockType`](crate::types::BlockType) could not be parsed.
    /// - Contains `None` if the end of input was unexpectedly encountered.
    /// - Contains `Some` negative value if an unrecognized encoding for a type was encountered.
    /// - Contains `Some` positive value if the parsed [`TypeIdx`] was too large.
    ///
    /// [`TypeIdx`]: crate::module::TypeIdx
    BlockType(Option<core::num::NonZeroI64>),
    /// A [`ValType`](crate::types::ValType) was not valid.
    ///
    /// - Contains `None` if a [`BlockType::Empty`] was parsed.
    /// - Contains `Some` type index if a [`BlockType::Index`] was parsed.
    ///
    /// [`BlockType::Empty`]: crate::types::BlockType::Empty
    /// [`BlockType::Index`]: crate::types::BlockType::Index
    ValType(Option<crate::module::TypeIdx>),
    Limits {
        index_type: crate::types::IdxType,
        component: LimitsComponent,
    },
    RefType(crate::types::ValType),
    #[non_exhaustive]
    MemType,
    #[non_exhaustive]
    TableType,
    #[non_exhaustive]
    GlobalType,
    #[non_exhaustive]
    TagType,
    #[non_exhaustive]
    ImportDesc {
        kind: u8,
    },
    Import(ImportComponent),
    ModuleSectionOrder(crate::ordering::OrderingError<crate::module::ModuleSectionOrder>),
    Opcode(crate::isa::InvalidOpcode),
    #[non_exhaustive]
    Instr {
        opcode: crate::isa::Opcode,
        reason: crate::isa::InvalidInstr,
    },
    Expr(crate::isa::InvalidExpr),
    MemArg(MemArgComponent),
}

crate::static_assert::check_size!(ErrorCause, <= 16);

impl From<nom::error::ErrorKind> for ErrorCause {
    #[inline]
    fn from(error: nom::error::ErrorKind) -> Self {
        Self::Nom(error)
    }
}

impl ErrorCause {
    /// Attempts to map this [`ErrorCause`] to its closest [`nom::error::ErrorKind`] counterpart.
    ///
    /// This conversion may result in the loss of error information.
    pub fn to_error_kind(self) -> nom::error::ErrorKind {
        use crate::{
            isa::{InvalidExpr, InvalidInstr},
            values::InvalidVector,
        };
        use nom::error::ErrorKind as Kind;

        match self {
            Self::Nom(kind) => kind,
            Self::Leb128 { .. } | Self::Index(_) => Kind::ManyTill,
            Self::InvalidTag(_)
            | Self::PreambleMagic(_)
            | Self::PreambleVersion(_)
            | Self::ImportDesc { .. }
            | Self::Opcode(_) => Kind::Tag,
            Self::InvalidFlags(_) => Kind::OneOf,
            Self::Vector(InvalidVector::Length) | Self::NameLength | Self::SectionLength => {
                Kind::LengthValue
            }
            Self::Vector(InvalidVector::Remaining { .. }) => Kind::Count,
            Self::NameContents(_) | Self::SectionContents(_) => Kind::Complete,
            Self::NameEncoding(_)
            | Self::BlockType(_)
            | Self::ValType(_)
            | Self::RefType(_)
            | Self::ModuleSectionOrder(_)
            | Self::Expr(InvalidExpr::BlockNestingOverflow)
            | Self::Expr(InvalidExpr::ExpectedEnds(_))
            | Self::Instr {
                reason:
                    InvalidInstr::Unrecognized
                    | InvalidInstr::BrTableLabelCount
                    | InvalidInstr::SelectTypedArity(_),
                ..
            } => Kind::Verify,
            Self::SectionId
            | Self::CustomSectionName
            | Self::Limits { .. }
            | Self::MemType
            | Self::TableType
            | Self::GlobalType
            | Self::TagType
            | Self::Import(_)
            | Self::MemArg(_)
            | Self::Instr {
                reason:
                    InvalidInstr::Argument
                    | InvalidInstr::Source
                    | InvalidInstr::Destination
                    | InvalidInstr::VectorLane,
                ..
            } => Kind::Eof,
        }
    }
}

impl Display for ErrorCause {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Nom(err) => write!(f, "could not parse: {}", err.description()),
            Self::Leb128 {
                destination,
                reason,
            } => {
                use crate::values::leb128::{Destination, InvalidEncoding};

                f.write_str("could not parse LEB128 encoded ")?;

                match destination {
                    Destination::U32 | Destination::U64 => f.write_str("un")?,
                    Destination::S32 | Destination::S64 => (),
                }

                f.write_str("signed ")?;

                match destination {
                    Destination::U32 | Destination::S32 => f.write_str("32")?,
                    Destination::U64 | Destination::S64 => f.write_str("64")?,
                }

                f.write_str("-bit integer")?;

                match reason {
                    InvalidEncoding::Overflow => {
                        write!(f, ", an overflow occured while decoding the value")
                    }
                    InvalidEncoding::NoContinuation => Ok(()),
                }
            }
            Self::InvalidTag(tag) => Display::fmt(tag, f),
            Self::InvalidFlags(flags) => Display::fmt(flags, f),
            Self::Vector(bad) => Display::fmt(bad, f),
            Self::NameLength => f.write_str("expected name length"),
            Self::NameContents(e) => e.print("UTF-8 encoded name", f),
            Self::NameEncoding(e) => write!(f, "invalid name encoding: {e}"),
            Self::Index(name) => write!(f, "could not parse {name} index"),
            Self::SectionId => f.write_str("missing section ID byte"),
            Self::SectionLength => f.write_str("expected section content length"),
            Self::SectionContents(e) => e.print("section contents", f),
            Self::CustomSectionName => f.write_str("expected custom section name"),
            Self::PreambleMagic(bad) => Display::fmt(bad, f),
            Self::PreambleVersion(None) => f.write_str("missing WASM preamble version"),
            Self::PreambleVersion(Some(actual)) => {
                let expected = u32::from_le_bytes(crate::module::preamble::RECOGNIZED_VERSION);
                write!(f, "expected WASM preamble version {expected} ({expected:#010X}), but got {actual} ({actual:#010X})")
            }
            Self::BlockType(None) => f.write_str("expected valtype, typeidx, or empty block type"),
            Self::BlockType(Some(block_type)) => {
                if block_type.get() < 0 {
                    write!(f, "{block_type} is not a valid value type or block type")
                } else {
                    write!(f, "type index in block type {block_type} is too large, maximum 32-bit value is {}", u32::MAX)
                }
            }
            Self::ValType(None) => f.write_str("expected valtype but got empty block type"),
            Self::ValType(Some(index)) => write!(f, "expected valtype but got type index {index}"),
            Self::Limits {
                index_type,
                component,
            } => {
                f.write_str("could not parse ")?;
                f.write_str(match index_type {
                    crate::types::IdxType::I32 => "32",
                    crate::types::IdxType::I64 => "64",
                })?;
                write!(f, "-bit integer {component} bound for limit")
            }
            Self::RefType(actual) => write!(f, "expected reftype but got {actual}"),
            Self::MemType => f.write_str("could not parse memory type"),
            Self::TableType => f.write_str("could not parse table type"),
            Self::GlobalType => f.write_str("could not parse global type"),
            Self::TagType => f.write_str("could not parse tag type"),
            Self::ImportDesc { kind } => write!(f, "error parsing importdesc kind {kind:#04X}"),
            Self::Import(field) => write!(f, "could not parse import: missing {field}"),
            Self::ModuleSectionOrder(order) => Display::fmt(order, f),
            Self::Opcode(bad) => Display::fmt(bad, f),
            Self::Instr { opcode, reason } => {
                write!(f, "could not parse `{opcode}` instruction {reason}")
            }
            Self::Expr(bad) => Display::fmt(bad, f),
            Self::MemArg(bad) => write!(f, "could not parse memarg: {bad}"),
        }
    }
}

#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
#[cfg(feature = "std")]
impl std::error::Error for ErrorCause {
    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
        Some(match self {
            Self::InvalidTag(e) => e,
            Self::InvalidFlags(e) => e,
            Self::NameEncoding(e) => e,
            Self::PreambleMagic(e) => e,
            Self::ModuleSectionOrder(e) => e,
            Self::Opcode(e) => e,
            Self::Instr { reason, .. } => reason,
            Self::Expr(e) => e,
            _ => return None,
        })
    }
}
