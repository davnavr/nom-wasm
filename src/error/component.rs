use core::fmt::{Display, Formatter};

/// Indicates which part of a [`Limits`](crate::types::Limits) could not be parsed.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[non_exhaustive]
#[allow(missing_docs)]
pub enum LimitsComponent {
    Minimum,
    Maximum,
}

impl Display for LimitsComponent {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.write_str(match self {
            Self::Minimum => "minimum",
            Self::Maximum => "maximum",
        })
    }
}

/// Indicates which field of an [`Import`](crate::module::Import) could not be parsed.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[non_exhaustive]
#[allow(missing_docs)]
pub enum ImportComponent {
    Module,
    Name,
}

impl Display for ImportComponent {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        f.write_str(match self {
            Self::Module => "module name",
            Self::Name => "import name",
        })
    }
}

/// Indicates why a [`MemArg`](crate::isa::MemArg) could not be parsed.
#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
#[non_exhaustive]
#[allow(missing_docs)]
pub enum MemArgComponent {
    /// Indicates that the [**`align`**] field could not be parsed when [`None`]; otherwise,
    /// indicates that the [**`align`**] was too large.
    ///
    /// [**`align`**]: crate::isa::MemArg::align
    Alignment(Option<u32>),
    Offset,
    Memory,
}

impl Display for MemArgComponent {
    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Alignment(Some(a)) => {
                write!(f, "specified alignment was 2^{a}, which is too large")
            }
            Self::Alignment(None) => f.write_str("alignment field"),
            Self::Offset => f.write_str("offset field"),
            Self::Memory => f.write_str("memory field"),
        }
    }
}
