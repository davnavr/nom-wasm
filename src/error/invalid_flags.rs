/// Used with [`InvalidFlags`] to indicate what values were invalid.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[allow(clippy::exhaustive_enums)]
pub enum InvalidFlagsValue<V: Copy> {
    /// The flags were parsed, but had invalid flags.
    Invalid {
        /// The flags value that *contains* the invalid flags.
        value: V,
        /// The flags that caused validation to fail.
        invalid: V,
    },
    /// Flags could not be parsed.
    Missing,
}

/// Error type used when a byte or 32-bit flags combination was invalid.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[non_exhaustive]
pub enum InvalidFlags {
    /// Invalid flags for [`Limits`](crate::types::Limits).
    Limits(InvalidFlagsValue<u8>),
    /// Invalid flags for a [`GlobalType`](crate::types::GlobalType).
    GlobalType(InvalidFlagsValue<u8>),
}

impl core::fmt::Display for InvalidFlags {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        let (name, invalid) = match self {
            Self::Limits(e) => ("limits", e),
            Self::GlobalType(e) => ("global type", e),
        };

        match invalid {
            InvalidFlagsValue::Invalid { value, invalid } => write!(
                f,
                "the {name} flags {value:#04X} contains invalid flag(s): {invalid:#04X}"
            ),
            InvalidFlagsValue::Missing => write!(f, "missing {name} flags"),
        }
    }
}

#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
#[cfg(feature = "std")]
impl std::error::Error for InvalidFlags {}
