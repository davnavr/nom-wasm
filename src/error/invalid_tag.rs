/// Error type used when a byte or 32-bit enumeration value was invalid.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[non_exhaustive]
pub enum InvalidTag {
    /// An invalid [`ModuleSectionId`](crate::module::ModuleSectionId).
    ModuleSectionId(u8),
    #[allow(missing_docs)]
    FuncType(Option<u8>),
    /// An invalid [`ImportDesc`](crate::module::ImportDesc).
    ImportDesc(Option<u8>),
}

impl core::fmt::Display for InvalidTag {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        let (value, value_width) = match self {
            Self::ModuleSectionId(b) => (Some(u32::from(*b)), 4),
            Self::FuncType(b) | Self::ImportDesc(b) => (b.map(u32::from), 4),
        };

        let name = match self {
            Self::ModuleSectionId(_) => "module section ID",
            Self::FuncType(_) => "function type",
            Self::ImportDesc(_) => "import desc",
        };

        if let Some(value) = value {
            write!(
                f,
                "the {name} tag {value:#0value_width$X} ({value}) is invalid"
            )
        } else {
            write!(f, "missing {name} tag")
        }
    }
}

#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
#[cfg(feature = "std")]
impl std::error::Error for InvalidTag {}
