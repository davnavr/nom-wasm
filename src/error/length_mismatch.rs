/// Describes an [`ErrorCause`] where the length of some data was incorrect.
///
/// [`ErrorCause`]: crate::error::ErrorCause
#[derive(Clone, Debug, Eq, PartialEq)]
#[allow(missing_docs)]
pub struct LengthMismatch {
    pub expected: u32,
    pub actual: u32,
}

impl LengthMismatch {
    pub(in crate::error) fn print(
        &self,
        name: &str,
        f: &mut core::fmt::Formatter,
    ) -> core::fmt::Result {
        write!(
            f,
            "expected {} bytes for {name}, but got {}",
            self.expected, self.actual
        )
    }
}

impl core::fmt::Display for LengthMismatch {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        write!(
            f,
            "expected {} bytes but got {}",
            self.expected, self.actual
        )
    }
}

#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
#[cfg(feature = "std")]
impl std::error::Error for LengthMismatch {}
