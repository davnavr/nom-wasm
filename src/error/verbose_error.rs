use crate::error;
use alloc::boxed::Box;
use core::fmt::Debug;
use nom::error::ParseError;

#[derive(PartialEq)]
enum Error<'a> {
    Error(nom::error::Error<&'a [u8]>),
    Cause(error::ErrorCause),
}

impl Debug for Error<'_> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        match self {
            Self::Error(e) => Debug::fmt(e, f),
            Self::Cause(c) => Debug::fmt(c, f),
        }
    }
}

#[derive(PartialEq)]
struct Inner<'a> {
    base: nom::error::Error<&'a [u8]>,
    additional: alloc::vec::Vec<Error<'a>>,
}

/// Accumulates information about an error, including its location, the kinds of error that
/// occured, and the reasons why it occured.
#[derive(PartialEq)]
#[repr(transparent)]
pub struct VerboseError<'a> {
    inner: Box<Inner<'a>>,
}

impl Debug for VerboseError<'_> {
    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
        f.debug_list()
            .entry(&self.inner.base)
            .entries(&self.inner.additional)
            .finish()
    }
}

impl<'a> ParseError<&'a [u8]> for VerboseError<'a> {
    fn from_error_kind(input: &'a [u8], kind: nom::error::ErrorKind) -> Self {
        Self {
            inner: Box::new(Inner {
                base: ParseError::from_error_kind(input, kind),
                additional: Default::default(),
            }),
        }
    }

    fn append(input: &'a [u8], kind: nom::error::ErrorKind, mut other: Self) -> Self {
        other
            .inner
            .additional
            .push(Error::Error(ParseError::from_error_kind(input, kind)));

        other
    }
}

impl<'a> error::ErrorSource<'a> for VerboseError<'a> {
    fn from_error_kind_and_cause(
        input: &'a [u8],
        kind: nom::error::ErrorKind,
        cause: error::ErrorCause,
    ) -> Self {
        Self {
            inner: Box::new(Inner {
                base: ParseError::from_error_kind(input, kind),
                additional: vec![Error::Cause(cause)],
            }),
        }
    }

    fn with_cause(mut self, cause: error::ErrorCause) -> Self {
        self.inner.additional.push(Error::Cause(cause));
        self
    }
}
