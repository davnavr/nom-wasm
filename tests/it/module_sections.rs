use nom_wasm::error::VerboseError;

/*
#[test]
fn type_sec_example() {
    let bytes = Bytes::from(&[
        1u8,  // count
        0x60, // func
        1,    // parameter count
        0x7F, // i32
        1,    // result count
        0x7E, // i64
    ]);

    let mut errors = InlineErrorReporter::<8>::new();
    let result = parser::Vector::with_parsed_count(&mut errors, bytes)
        .map(module::TypeSec::from)
        .map_err(|_| errors.expect_report());

    insta::assert_debug_snapshot!(result)
}
*/

#[test]
fn import_sec_example() {
    let mut bytes = Vec::with_capacity(128);
    bytes.extend([
        4, // count
        3, // module name length
    ]);
    bytes.extend(b"env");
    bytes.push(0xB); // name length
    bytes.extend(b"doSomeStuff");
    bytes.extend([
        0, // import func
        0, // typeidx
        3, // module name length
    ]);
    bytes.extend(b"env");
    bytes.push(6); // name length
    bytes.extend(b"memory");
    bytes.extend([
        2,    // import memory
        0,    // limit w/o maximum
        0x10, // limit minimum
        2,    // module name length
    ]);
    bytes.extend(b"rt");
    bytes.push(0xA); // name length
    bytes.extend(b"references");
    bytes.extend([
        1,    // import table,
        0x6F, // externref
        0, 0, // limits
        2, // module name length
    ]);
    bytes.extend(b"rt");
    bytes.push(8); // name length
    bytes.extend(b"stackptr");
    bytes.extend([
        3,    // import global
        0x7F, // i32
        1,    // mutable
    ]);

    let result = nom_wasm::module::ImportSec::parse::<VerboseError>(&bytes);

    insta::assert_debug_snapshot!(result);
}
