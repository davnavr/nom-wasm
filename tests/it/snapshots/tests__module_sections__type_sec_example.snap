---
source: tests/it/module_sections.rs
expression: result
---
Ok(
    [
        FuncType {
            parameters: [
                I32,
            ],
            results: [
                I64,
            ],
        },
    ],
)
